components/Common/CodeBox/index.tsx

```diff
@@ -0,0 +1,119 @@
+import {
+  DocumentDuplicateIcon,
+  CodeBracketIcon,
+} from '@heroicons/react/24/outline';
+import { useTranslations } from 'next-intl';
+import type { FC, PropsWithChildren, ReactNode } from 'react';
+import { Fragment, isValidElement, useMemo, useRef } from 'react';
+
+import Button from '@/components/Common/Button';
+import { useCopyToClipboard, useNotification } from '@/hooks';
+import { getCodeLanguageDisplayName } from '@/util/getCodeLanguageDisplayName';
+
+import styles from './index.module.css';
+
+type CodeBoxProps = {
+  language: string;
+};
+
+export const CodeBox: FC<PropsWithChildren<CodeBoxProps>> = ({
+  children,
+  language,
+}) => {
+  const ref = useRef<HTMLPreElement>(null);
+
+  const notify = useNotification();
+  const [, copyToClipboard] = useCopyToClipboard();
+  const t = useTranslations();
+
+  const onCopy = () => {
+    if (ref.current?.textContent) {
+      copyToClipboard(ref.current.textContent);
+
+      notify({
+        duration: 3000,
+        message: (
+          <div className={styles.notification}>
+            <CodeBracketIcon className={styles.icon} />
+            {t('components.common.codebox.copied')}
+          </div>
+        ),
+      });
+    }
+  };
+
+  const code = useMemo(() => transformCode(children), [children]);
```

Comment:
Is memoization needed here? I wonder how often this CodeBox would get re-rendered? Afaik the children and language props would never change.

Hence, no memorization is needed here

Reply:
Makes sense.